libcaf_net/caf/net/http/client_factory.hpp

```diff
@@ -0,0 +1,174 @@
+    return request(http::method::get);
+  }
+
+  /// Sends a HTTP POST message.
+  expected<std::pair<async::future<response>, disposable>>
+  post(std::string_view payload) {
+    return request(http::method::post, payload);
+  }
+
+  /// Sends a HTTP PUT message.
+  expected<std::pair<async::future<response>, disposable>>
+  put(std::string_view payload) {
+    return request(http::method::post, payload);
+  }
+
+  /// Sends a HTTP DELETE message.
+  expected<std::pair<async::future<response>, disposable>> del() {
+    return request(http::method::get);
+  }
+
+  /// Sends a HTTP CONNECT message.
+  expected<std::pair<async::future<response>, disposable>> connect() {
+    return request(http::method::get);
+  }
+
+  /// Sends a HTTP OPTIONS message.
+  expected<std::pair<async::future<response>, disposable>>
+  options(std::string_view payload) {
+    return request(http::method::post, payload);
+  }
+
+  /// Sends a HTTP TRACE message.
+  expected<std::pair<async::future<response>, disposable>>
+  trace(std::string_view payload) {
+    return request(http::method::post, payload);
+  }
+
+  /// Utility function to make a request with given parameters.
+  expected<std::pair<async::future<response>, disposable>>
+  request(http::method method, std::string_view payload = std::string_view{}) {
+    // Only connecting to an URI is enabled in the 'with' DSL.
+    auto& cfg = super::config();
+    CAF_ASSERT(std::holds_alternative<dsl::client_config::lazy>(cfg.data)
+               || std::holds_alternative<caf::error>(cfg.data));
+    if (std::holds_alternative<caf::error>(cfg.data))
+      return do_start(super::config(), std::get<caf::error>(cfg.data));
+    auto& data = std::get<dsl::client_config::lazy>(cfg.data);
+    CAF_ASSERT(std::holds_alternative<uri>(data.server));
+    const auto& resource = std::get<uri>(data.server);
+    cfg.method = method;
+    cfg.path = resource.path_query_fragment();
+    cfg.payload = payload;
+    cfg.fields.insert(
+      std::pair{std::string{"Host"}, resource.authority().host_str()});
+    return do_start(super::config(), data);
+  }
+
+private:
+  using return_t = expected<std::pair<async::future<response>, disposable>>;
+
+  template <typename Conn>
+  return_t do_start_impl(config_type& cfg, Conn conn) {
+    using transport_t = typename Conn::transport_type;
+    auto app_t = async_client::make(cfg.method, cfg.path, cfg.fields,
+                                    cfg.payload);
+    auto ret = app_t->get_future();
+    auto http_client = http::client::make(std::move(app_t));
+    auto transport = transport_t::make(std::move(conn), std::move(http_client));
+    transport->active_policy().connect();
+    auto ptr = net::socket_manager::make(cfg.mpx, std::move(transport));
+    cfg.mpx->start(ptr);
+    return std::pair{std::move(ret), disposable{std::move(ptr)}};
+  }
+
+  return_t do_start(config_type& cfg, dsl::client_config::lazy& data) {
```

Comment:
There's quite a bit of code in there. We're not in a template here, so why not put this into a `.cpp` file? Probably also applies to `request`.

Reply:
Separated.